cmd/schema.go

```diff
@@ -0,0 +1,85 @@
+package cmd
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"path"
+	"time"
+
+	"github.com/cloudspannerecosystem/harbourbridge/common"
+	"github.com/cloudspannerecosystem/harbourbridge/conversion"
+	"github.com/cloudspannerecosystem/harbourbridge/internal"
+	"github.com/google/subcommands"
+)
+
+// SchemaCmd struct with flags.
+type SchemaCmd struct {
+	filePrefix       string
+	driverName       string
+	schemaSampleSize int64
+	dumpFilePath     string
+	targetDb         string
+}
+
+// Name returns the name of operation.
+func (cmd *SchemaCmd) Name() string {
+	return "schema"
+}
+
+// Synopsis returns summary of operation.
+func (cmd *SchemaCmd) Synopsis() string {
+	return "generate schema for target db from source db schema"
+}
+
+// Usage returns usage info of the command.
+func (cmd *SchemaCmd) Usage() string {
+	return fmt.Sprintf(`%v schema [file]...
+
+Convert schema for source db specified by driver. Source db dump file can be
+specified by either dump-file flag or piped to stdin. Connection profile for source
+databases in direct connect mode can be specified by setting appropriate
+environment variables. The schema flags are:
+`, path.Base(os.Args[0]))
+}
+
+// SetFlags sets the flags.
+func (cmd *SchemaCmd) SetFlags(f *flag.FlagSet) {
+	f.StringVar(&cmd.filePrefix, "prefix", "", "File prefix for generated files")
+	f.StringVar(&cmd.driverName, "driver", "pg_dump", "Flag for specifying source DB or dump files (accepted values are \"pg_dump\", \"postgres\", \"mysqldump\", \"mysql\", and \"dynamodb\")")
+	f.Int64Var(&cmd.schemaSampleSize, "schema-sample-size", int64(100000), "Number of rows to use for inferring schema (only for DynamoDB)")
+	f.StringVar(&cmd.dumpFilePath, "dump-file", "", "Location of dump file to process")
+	f.StringVar(&cmd.targetDb, "target-db", "", "Specifies the target DB, defaults to spanner")
+}
+
+func (cmd *SchemaCmd) Execute(ctx context.Context, f *flag.FlagSet, _ ...interface{}) subcommands.ExitStatus {
+	input := common.OpenDump(cmd.dumpFilePath)
```

Comment:
It's a bit confusing to call OpenDump when we might not have a dump file or maybe we are using a direct-connect driver, so we don't even need the dump file.
(This was also an issue with the previous code with the loadInput function, which was even less meaningful).

I don't have a good resolution for this.

I can't think of a way to improve the naming of OpenDump -- OptionallyOpenDump or ConfigureInput seem worse ...

And I can't think of a good comment to add. Something like:

// OpenDump file it is is specified.

seems pretty worthless.

Reply:
I thought about it a bit and realized that OpenDump/LoadInput wasn't adding much meaning whereas, what we really wanted was to get reference to a new IOStreams struct depending on whether we are in dump mode etc. So refactored out the code to create the following helper function.

```
// NewIOStreams returns a new IOStreams struct such that input stream is set
// to open file descriptor for dumpFile if driver is PGDUMP or MYSQLDUMP.
// Input stream defaults to stdin. Output stream is always set to stdout.
func NewIOStreams(driver string, dumpFile string) IOStreams {
	io := IOStreams{In: os.Stdin, Out: os.Stdout}
	if (driver == PGDUMP || driver == MYSQLDUMP) && dumpFile != "" {
		fmt.Printf("\nLoading dump file from path: %s\n", dumpFile)
		f, err := os.Open(dumpFile)
		if err != nil {
			fmt.Printf("\nError reading dump file: %v err:%v\n", dumpFile, err)
			log.Fatal(err)
		}
		io.In = f
	}
	return io
}
```

Let me know what do you think?

Reply:
Yes, I think that is much better.
Thanks for thinking about this and coming up with a good solution.